### DISCUSSION TITLE
Will there be a first-party way of compling .NET to WebAssembly in .NET 7 without Blazor?

### DISCUSSION THREAD

**Discussion by torgabor** [dotnet/runtime#71122] [General] [answered]
Hi! 
Sorry, if this has been answered, but I haven't been able to find anything conclusive on this.
I want to use my .NET code from Javascript so that they look like native Js objects from the outside, but I don't want to include Blazor, since I'm not really interested in interacting with the DOM from .NET.

There are a number of third-party solutions to do this, such as [Uno](https://github.com/unoplatform/Uno.Wasm.Bootstrap) or [DotnetJS](https://github.com/Elringus/DotNetJS), both of which seem to be doing exactly what I'm looking for. 

My question is, is there or will there be an 'official' alternative for this?

Sorry if this question is kind of redundant,  but I couldn't find anything so far :)

**Comment by maraf** [comment]
Hello @torgabor,

Similar question arrived few days before https://github.com/dotnet/runtime/discussions/70632.

You can try out [templates](https://www.nuget.org/packages/Microsoft.NET.Runtime.WebAssembly.Templates/7.0.0-preview.5.22301.12) or take a look at [samples](https://github.com/dotnet/runtime/tree/main/src/mono/sample/wasm). There is an ongoing work make this scenario better.

**Comment by torgabor** [comment]
Thanks for the answer, this the kind of thing I was looking for. I've looked through the examples, and they seem to be focused on invoking static methods in .NET. Is there an equivalent of `DotNetObjectReference` and `IJSRuntime` that would allow me to access instance methods, or callback JS code from .NET?

**Comment by maraf** [comment, marked as answer]
Actually, a new interop API is being developed. You can take a look at the proposal https://github.com/dotnet/runtime/issues/70133 and prototype implementation https://github.com/dotnet/runtime/pull/66304.

The API will be different to what `IJSRuntime` offers, it is more low level and oriented on allocation and performance. And yes, you can call instance methods.

Take a look at the API and let me know if you have more questions. I'll be happy to answer them.

**Reply by torgabor** [reply]
Thanks for the answer! It seems it's still a work in progress. I'll revisit this problem once the issues you link are resolved.
